Setup dual Redis store for rolling over fragment cacheing to Redis

mstruve opened the pull request:
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
This new cache store, `dual_rails_store` will allow us to continue to write to Memcache while moving over keys to Redis at our own pace. I modeled this class off of a [gem I found for dual caches](https://github.com/mezis/level2) except I updated it for our purposes. 

* Any cache statement as-is will use Memcache 
* Any cache statement with the option `only: redis` will only use Redis
* Any cache statement with the option `all: true` will write to Redis and Memcache but will read/return the value read from Memcache.

What does everyone think of this approach? I like that we can continue to roll over keys at our own pace and that we can back populate any keys we deem a higher risk to roll over. 

## Related Tickets & Documents
#4670 

## Added to documentation?
- [x] no documentation needed

![maybe that will work gif](https://media1.tenor.com/images/63f988e6b715c4d94b830b07d9ecb396/tenor.gif?itemid=13407671)

rhymes commented on config/environments/development.rb:
```diff
@@ -1,3 +1,5 @@
+require 'active_support/cache/dual_rails_store'
```
I don't think we need this because `lib` is preloaded

rhymes commented on lib/active_support/cache/dual_rails_store.rb:
```diff
@@ -0,0 +1,71 @@
+require 'active_support/cache'
+require 'active_support/version'
+
+module ActiveSupport
```
I would add all the explanation you added in the PR as a comment here, because even if this is temporary I think having an explanation in the code that survives the PR life will make it easier for people that come across to this file to understand what it's supposed to do.

I would also mention it's adapted from <https://github.com/mezis/level2>

mstruve commented:
Why do you have to be so responsible @rhymes 😉

mstruve commented:
Thanks for the thorough investigation @rhymes I have my own little memcache setup in Heroku, I will point it at that and dig into some of these.

mstruve commented:
Definitely looking into these two right now
1. the dual store doesn't work correctly with Dalli
2. default expiration nil

I don't think this is a problem, we should never be setting expire to 0 anyways.
1. expires_in: 0 breaks with Redis

mstruve commented:
Turns out `dalli_store` overrides quite a few of the methods that I put into place with my new `DualRailsStore` which is why this breaks so badly with it. 

[read_entry](https://github.com/petergoldstein/dalli/blob/v2.7.8/lib/active_support/cache/dalli_store.rb#L301) is overridden to only return the value rather than the active support cache entry object. This looks like its due to maintaining backwards compatibility for something. I updated my read_entry method to ensure that a valid entry object is always returned and that seem to fix all of the issues and errors I was seeing. 

[write_entry](https://github.com/petergoldstein/dalli/blob/v2.7.8/lib/active_support/cache/dalli_store.rb#L313) is also set by `dalli_store` and from looking at the override we can see that we need to add a connection object to the options hash in order for it to work. This is why I added a line to do that in my `write_entry` method.

expiration - The default expiration was getting lost in the shuffle so I made sure to add it to any options hashes that come in with expires_at nil.

rhymes commented:
@mstruve woah, a lot of patching around :( 

It seems to be working for me, I re-tested all the cases (except the one with `expires_in` set to 0). The ttl is set correctly in all cases and this works correctly with memcached:

```ruby
[1] pry(main)> Rails.cache.write("foo4", "bar", all: true)
Dalli::Server#connect 127.0.0.1:11211
=> "OK"
```

There's something weird that I found out by mistake:

```ruby
[19] pry(main)> Rails.cache.write("foobar", "xyz", only: :memcached)
=> "OK"
[20] pry(main)> Rails.cache.stores[:default_store].read("foobar")
Cache read: foobar
=> nil
[21] pry(main)> Rails.cache.stores[:redis].read("foobar")
=> nil
```

If the value of only is unknown the calls works, but the data is not store anywhere. Should we raise an error if the store is not known?

rhymes commented on spec/lib/active_support/cache/dual_rails_store_spec.rb:
```diff
@@ -54,7 +54,7 @@
 
     it "honors default expiration" do
-      dual_store.write("foo", "bar")
+      dual_store.write("foo", "bar", all: true)
```
we might want to check the TTL in both stores then...

benhalpern commented:
Good for deployment IMO @mstruve
